Fix for CondFD Problem that Improperly Led to Fatal Error

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #9597
 - User had an input file that requested CondFD for a single construction that was leading to a fatal error crash.  The problem was caused by an input file that used regular materials to define air gaps and then asked for one construction to use the CondFD solution (that didn't actually use that regular material defined air gap).  The CondFD code was set up to establish arrays for all constructions except a couple scenarios.  This caused it to try to initialize data for a construction that was not even using the CondFD solution and led to problems in the initialization routine (i.e., a fatal error to avoid a potentially unstable situation).  The problem went away when the material was defined using Material:AirGap because the code had logic for that scenario while is entirely appropriate.  However, there was no reason for the CondFD algorithm to go constructions that did not use the CondFD solution.  In most cases, it didn't cause a problem.  In this case, it threw an inappropriate fatal error.  The fix involves a new subroutine that checks to see if a particular construction actually uses CondFD and skipping the ones that don't.  This allowed the user's input file that caused the problem in the first place to run to completion as expected.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Correct CondFD Error

User had an input file where an air layer was defined with a regular material and then a construction that did not use that layer used the CondFD solution.  CondFD looked at everything and then produced an error message because the air layer didn't fit the CondFD requirements.  This has now been fixed and CondFD will only look at constructions that actually use the CondFD solution.  A new subroutine and unit test are included.

RKStrand committed:
Merge branch 'develop' into 9597SettingConstrToCondFDFatalError

RKStrand committed:
Merge branch 'develop' into 9597SettingConstrToCondFDFatalError

rraustad commented on src/EnergyPlus/HeatBalFiniteDiffManager.cc:
```diff
@@ -2626,6 +2629,17 @@ namespace HeatBalFiniteDiffManager {
         updatedThermalConductivity = materialDefinition.phaseChange->getConductivity(temperatureUpdated);
     }
 
+    bool findAnySurfacesUsingConstructionAndCondFD(EnergyPlusData &state, int constructionNum)
```
int const contructionNum, since it's not changing.

RKStrand replied:
Yup.  This was an oversight on my part.  I've fixed it in the latest commit.

RKStrand committed:
FiniteDiffManager const Argument

Correction of variable declaration in arguments of new function.  Variable was a constant but not declared as such incorrectly.

RKStrand committed:
Merge branch 'develop' into 9597SettingConstrToCondFDFatalError

rraustad commented on src/EnergyPlus/HeatBalFiniteDiffManager.cc:
```diff
@@ -2626,6 +2629,17 @@ namespace HeatBalFiniteDiffManager {
+            if (thisSurface.Construction == constructionNum) {
+                if (thisSurface.HeatTransferAlgorithm == DataSurfaces::HeatTransferModel::CondFD) return true;
+            }
+        }
```
for (auto &thisSurface : state.dataSurface->Surface) {
            if (thisSurface.Construction == constructionNum) {
                if (thisSurface.HeatTransferAlgorithm == DataSurfaces::HeatTransferModel::CondFD) return true;
            }
        }

Update: use auto &

RKStrand replied:
Ok--made this change as suggested and pushed the commit.

rraustad commented:
@RKStrand the defect file AUE_BS from issue 9597 runs on Windows in develop. I added Output:Constructions,Contructions,Materials; to the defect file and the file still runs in develop. Is there something else that needs to be added to the defect file? I am trying to hit your new function in this branch in the debugger (to verify my suggested code change) and have yet to do so.

RKStrand commented:
@rraustad I think that the unmet hours post clarifies what change you have to make to get the problem.  One of the surfaces in the file has to be changed to use CondFD.  The file itself uses CTFs but when you switch one surface to CondFD, you should see the problem happen.

RKStrand committed:
Update based on Review Comment

@rraustad suggested this simplification for this code.

rraustad commented:
OK, the defect file was changed as:

    !-   ===========  ALL OBJECTS IN CLASS: HEATBALANCEALGORITHM ===========
    HeatBalanceAlgorithm,ConductionTransferFunction,200;

    SurfaceProperty:HeatTransferAlgorithm:Construction,
      BO05Override,            !- Name
      ConductionFiniteDifference,  !- Algorithm          <-- changed in defect file
      BO05;                    !- Construction Name

    SurfaceProperty:HeatTransferAlgorithm:Construction,
      BO05ReversedOverride,    !- Name
      ConductionFiniteDifference,  !- Algorithm         <-- changed in defect file
      BO05 Reversed;           !- Construction Name


Now this defect file is showing fatal errors in develop for material layers. Changing the material thickness as the message suggests leads to another material layer with this same fatal warning.

    ** Severe  ** InitialInitHeatBalFiniteDiff: Found Material that is too thin and/or too highly conductive, material name = LUFTSCHICHT_9CM
    **   ~~~   ** High conductivity Material layers are not well supported by Conduction Finite Difference, material conductivity = 0.630 [W/m-K]
    **   ~~~   ** Material thermal diffusivity = 5.122E-004 [m2/s]
    **   ~~~   ** Material with this thermal diffusivity should have thickness > 0.52591 [m]
    **  Fatal  ** Preceding conditions cause termination.

This branch runs the defect file successfully.

    ************* EnergyPlus Completed Successfully-- 124 Warning; 0 Severe Errors; Elapsed Time=00hr 16min 11.21sec

rraustad commented:
Review is complete, just waiting on CI to finish and check for diffs and performance impact. This should merge shortly.

rraustad commented:
There is 1 eio diff in 1ZoneUncontrolledFourAlgorithms that needs to be explained. This diff is basically saying that there is information that is now not being reported. @RKStrand can you explain that?

![image](https://user-images.githubusercontent.com/5778156/189461666-bd32b68f-f808-4655-aca9-3ea330737057.png)

rraustad commented:
CI is clean with a single time out on DirectIndirectEvapCoolersVSAS.

RKStrand commented:
@rraustad Yes, this is actually correct because in develop it forces CondFD to do calculates for every surface even if only one uses the CondFD solution.  As a result, it produces a bunch of output that is nonsense because it isn't even used.  In this modification, it no longer does those calculations so this junk output does not appear anymore.  So, this is actually correct that it does not produce this output (it's actually an additional sign that the fix is doing what it is supposed to).  Does that help?

RKStrand commented:
I think I set a record for the number of times that the word "actually" was used in a single GitHub comment 🤦‍♂️
